WorkloadManagementSystem/DB/PilotsLoggingDB.py

```diff
@@ -0,0 +1,175 @@
+                                 pool_recycle = 3600, echo_pool = True )
+    self.sqlalchemySession = scoped_session( sessionmaker( bind = self.engine ) )
+    self.inspector = Inspector.from_engine( self.engine )
+
+  ##########################################################################################
+  def __initializeDB( self ):
+
+    tablesInDB = self.inspector.get_table_names( )
+
+    if not 'PilotsLogging' in tablesInDB:
+      try:
+        PilotsLogging.__table__.create( self.engine )
+      except SQLAlchemyError as e:
+        return S_ERROR(e)
+    else:
+      gLogger.debug( "Table PilotsLogging exists" )
+
+    return S_OK( )
+
+  ##########################################################################################
+  def addPilotsLogging( self, pilotUUID, timestamp, source, phase, status, messageContent ):
+    """Add new pilot logging entry"""
+
+    session = self.sqlalchemySession( )
+    logging = PilotsLogging( pilotUUID, timestamp, source, phase, status, messageContent )
+
+    try:
+      session.add( logging )
+    except SQLAlchemyError as e:
+      session.rollback( )
+      session.close( )
+      return S_ERROR( DErrno.ESQLA, "Failed to add PilotsLogging: " + e.message )
+
+    try:
+      session.commit( )
+    except SQLAlchemyError as e:
+      return S_ERROR("Failed to commit PilotsLogging: " + e.message)
+      session.rollback( )
+      session.close( )
+
+    return S_OK( )
+
+  ##########################################################################################
+  def getPilotsLogging( self, pilotUUID ):
+    """Get list of logging entries for pilot"""
+
+    session = self.sqlalchemySession( )
+
+    pilotLogging = []
+    for pl in session.query( PilotsLogging ).filter(PilotsLogging.pilotUUID == pilotUUID ).order_by(
+        PilotsLogging.timestamp ).all( ):
+      entry = {}
+      entry['pilotUUID'] = pl.pilotUUID
+      entry['timestamp'] = pl.timestamp
+      entry['source'] = pl.source
+      entry['phase'] = pl.phase
+      entry['status'] = pl.status
+      entry['messageContent'] = pl.messageContent
+      pilotLogging.append( entry )
+
+    return S_OK( pilotLogging )
+
+  ##########################################################################################
+  def deletePilotsLogging( self, pilotUUID ):
+    """Delete all logging entries for pilot"""
+
+    if isinstance( pilotUUID, basestring ):
+      pilotUUID = [pilotUUID, ]
+
+    session = self.sqlalchemySession( )
+
+    session.query( PilotsLogging ).filter( PilotsLogging.pilotUUID.in_( pilotUUID ) ).delete(
+      synchronize_session = 'fetch' )
+
+    try:
+      session.commit( )
+    except SQLAlchemyError as e:
+      return S_ERROR("Failed to commit: " + e.message)
+      session.rollback( )
+      session.close( )
+
+    return S_OK( )
+
+  ##########################################################################################
+
+class PilotsLogging( Base ):
```

Comment:
Maybe this is just my taste, but shouldn't this be defined at the beginning of the module?

Reply:
The main class in the file is PilotsLoggingDB (matching file name) that exposes interface for handling database, this one is just a helper/mapper.